src/compose.ts

```diff
@@ -346,3 +366,99 @@ function normalizeKeyValuePairs(
 		.fromPairs()
 		.value();
 }
+
+/**
+ * Parses a composition and returns a list of image descriptors that can
+ * be used to pull or build a service image. The given composition version
+ * must be equal to `DEFAULT_SCHEMA_VERSION`, or an exception is thrown.
+ * Normalise the composition before passing it to this function.
+ */
+async function expandContext(
+	c: Composition,
+	fileResolverCb: (path: string) => Promise<ReadStream>,
+): Promise<Composition> {
+	const cachedEnvironmentFiles: Dict<Dict<string>> = {};
+
+	// Apply all read env_files content to the services referncing the env_files
+	for (const service of Object.values(c.services)) {
+		let envFilePaths = service.env_file;
+		if (envFilePaths !== undefined) {
+			service.environment = service.environment ?? {};
+			if (!Array.isArray(envFilePaths)) {
+				envFilePaths = [envFilePaths];
+			}
+			for (const envFilePath of envFilePaths) {
+				const validatedPath = validateAndNormalizeFilePath(envFilePath);
```

Comment:
Can `validateAndNormalizeFilePath` by moved to `normalizeObjectToComposition`? That could make this function a little more straight forward.

Reply:
As I understand, the normalized composition will be stored as part of the target state. I think it's good for debugging purpose to store the validated and normalized path in there. Nonetheless, I don't know if this will lead to confusion of the customers, who may compare they docker-compose file and the target-state and may see differences.

Thus, I decided to not change the persisted env_file paths in the composition.

I'd like to have @dfunckt in the loop for this

Reply:
I think we're doing enough transformations to the original composition that it's okay to consider it a totally new representation of the input. Given that, I'd just drop the `env_file` attributes completely from the normalized composition, so that there can be no confusion about whether that step (loading and applying env_files) has been taken care of.

Reply:
@dfunckt I'll drop the env_file tag from the composition